DxDispatch/src/dxdispatch/Executor.cpp

```diff
@@ -158,18 +172,32 @@ void Executor::operator()(const Model::DispatchCommand& command)
             }
 
             PIXBeginEvent(m_device->GetCommandQueue(), PIX_COLOR(255, 255, 0), "Dispatch '%s'", command.dispatchableName.c_str());
-            
+
+            uint32_t timestampIndex = (iteration % timestampCount) * 2;
+
+            m_device->GetCommandList()->EndQuery(m_device->GetQueryHeap(), D3D12_QUERY_TYPE_TIMESTAMP, timestampIndex);
             dispatchable->Dispatch(command);
+            m_device->GetCommandList()->EndQuery(m_device->GetQueryHeap(), D3D12_QUERY_TYPE_TIMESTAMP, timestampIndex + 1);
+            m_device->GetCommandList()->ResolveQueryData(m_device->GetQueryHeap(), D3D12_QUERY_TYPE_TIMESTAMP, timestampIndex, 2, m_timestampResultBuffers, timestampIndex * sizeof(UINT64));
             
             if (dispatchableUsesDeviceCommandList)
             {
                 m_device->DispatchAndWait();
             }
 
+            if (iteration > 0 && iteration % timestampCount == 0) {
+                void* pData = nullptr;
+                D3D12_RANGE readRange = { 0, sizeof(UINT64) * 2 * timestampCount };
+                m_timestampResultBuffers->Map(0, &readRange, &pData);
+
+                const UINT64* pTimestamps = reinterpret_cast<UINT64*>(pData);
+                timestamps.insert(timestamps.end(), pTimestamps, pTimestamps + 2 * timestampCount);
+            }
```

Comment:
You can remove this from the loop if you resolve all the queries in one shot as I mentioned in previous comment. I think you'd be better off doing this:
- Inside the dispatch loop: EndQuery calls only.
- After the dispatch loop ends: resolve all query data, wait for the GPU to finish, then map the results buffer and fill CPU-side timestamp results.

Of course, you'd need to be careful when the user sets the `-t` option for fixed-duration execution, which sets iterations to UINT32_MAX. I think it's reasonable to allocate a query heap that has a limited size (e.g. `min(number of iterations * 2, 16384)`). If the num iterations exceeds what the query heap holds then start overwriting the earlier timestamps.

If you *really* want to capture timestamps for every possible iteration then you'd want to resolve & map the results only once you've run out of space in the query heap.

Reply:
I had the `-t` case in mind, so I wrote the whole system around copying-back the buffer as we fill it up. It's much easier to just assume the buffer is big enough / we don't care, and I guess this is the most typical use case. I made it allocate 16384 timestamps and just wrap around if we run out. This resolves a couple of other comments as well.